Add stub defintions

saulshanabrook opened the pull request:
In this PR I have started adding a `.pyi` stub file for the library, to expose the types statically for tools like MyPy to consume. I have started adding stubs to not only help future downstream users of this library but to also increase my familiarity to the features this library exposes.


* [Mypy docs on stubs](https://mypy.readthedocs.io/en/stable/stubs.html)
* [Py03 docs on adding stub files](https://pyo3.rs/v0.16.4/python_typing_hints.html#if-you-do-not-have-other-python-files)

I also added MyPy as a development dependency, so that we can use its `stubtest` tool to verify the stubs match the actual implementation.

I also added a test to github actions to verify the stubs.

TODO:
- [x] Add all classes and functions
- [x] Make tests run on PRs to verify in CI
- [x] FIx submodule issue

saulshanabrook committed:
Start adding type definitions and check in CI

saulshanabrook committed:
Add initial stubs and tests

saulshanabrook commented:
I have added some initial stubs that seem to work OK. I also added tests so that the test files are run through MyPy in CI. I had to rename a few things in the files for MyPy to pass. I also had to `# ignore` the `vars` calls, because it cannot verify statically that it will return a tuple instead of just one item.

Currently, the subtests are failing due to a `snake_egg.snake_egg` module which is present:

```
error: snake_egg.snake_egg failed to find stubs
Stub:
MISSING
Runtime:
<module 'snake_egg.snake_egg' from '/Users/saul/p/snake-egg/venv/lib/python3.9/site-packages/snake_egg/snake_egg.cpython-39-darwin.so'>

Found 1 error (checked 2 modules)
```

I am not sure why this module is being built. Maybe this is how `maturin` works by default, I haven't used it before. AFAIK you only need one top level `snake_egg` module instead of a submodule as well, which contains all the same functionality? Any advice here would be appreciated.

saulshanabrook committed:
Ignore missing stubs to ignore snake_egg.snake_egg mod

saulshanabrook commented:
Ah I see the submodule is [created by maturin](https://github.com/mhils/maturin/blob/main/guide/src/project_layout.md#pure-rust-project):

> Maturin will add a necessary `__init__.py` to the package when building the wheel.


I added `--ignore-missing-stubs` so that this error is now ignored.

saulshanabrook committed:
Run CI on pull requests

saulshanabrook commented:
This is ready for review now. If a maintainer approves the workflow, we can see if it passes in ci!

saulshanabrook commented on tests/math.py:
```diff
@@ -7,27 +7,26 @@
 #   + this disables some rules and tests
 # * the last three tests
 
-from snake_egg import EGraph, Rewrite, Var, vars
```
I realized these tests got formatted with [black](https://github.com/psf/black)  and [isort](https://github.com/pycqa/isort). If you prefer a more minimal diff, I can revert them back, and just keep the needed changes (added the `#type: ignore`s and renamed the `namedtuples`s)

mwillsey replied:
Yes, I think a more minimal diff would be good for now!

saulshanabrook replied:
Sonds good, I reverted them back to the originals and only added the minimal I could to make it pass with MyPy.

saulshanabrook committed:
Revert test python files back closer to master

saulshanabrook commented on tests/math.py:
```diff
@@ -10,24 +10,24 @@
 # Operations
-diff = namedtuple("Diff", "x y")
-inte = namedtuple("Integral", "x y")
+diff = namedtuple("Diff", "x y") # type: ignore
```
I reverted the rename of the named tuple types, instead opting to just ignore the error mypy raises that the variable name should match the name, to minimize the diff.

mwillsey replied:
Hmm. Can we either just capitalize this or ignore on a global level?

saulshanabrook replied:
👍 Sure I can refactor to capitalize them all

saulshanabrook commented on tests/math.py:
```diff
@@ -56,8 +56,8 @@ def eval_math(car, cdr):
-a, b, c, x, f, g, y = vars("a b c x f g y")
-list_rules = [
+a, b, c, x, f, g, y = vars("a b c x f g y") # type: ignore
+list_rules: List[List[Any]] = [
```
We need this type signature, otherwise mypy will complain later down when creating the rewrites because it doesn't know the inner types.

If we wanted to avoid this manual imprecise type here, we could instead just create the rewrite list directly instead of first making lists of items then creating it.

saulshanabrook committed:
Add Python 3.6 compat

Make typings compatible with python 3.6 by using typing-extensions
to backport @final

saulshanabrook commented:
The tests failed in CI previously because I was using the `@final` decorator from typing which was added in Python 3.8. The tests in CI run on Python 3.6, so I moved to import it from the `typing-extensions` package which backports static type imports to older Python version.

saulshanabrook committed:
Rename named tuples to remove type ignores

saulshanabrook commented:
@mwillsey I renamed all the named tuples to make them camel case to match the names, to remove the `# type: ignore` comments.

mwillsey commented:
Thanks!
